public/app/plugins/datasource/loki/querySplitting.ts

```diff
@@ -161,3 +158,38 @@ export function runPartitionedQuery(datasource: LokiDatasource, request: DataQue
 
   return response;
 }
+
+function getNextRequestPointers(requests: LokiGroupedRequest, requestIndex: number, requestN: number) {
+  // There's a pending metric request:
+  if (requestIndex === 0 && requests[1]?.partition[requestN - 1]) {
```

Comment:
i'd recommend we improve the types in this thing. there are certain assumptions that we have that are not there in the typescript types:
1. we assume the `requests` array's length is always `2`.
2. we assume the first item in the requests array is the logs-queries and the second item is the metric-queries.

so i wonder, if we could change the type a little to make it better, like:

```ts
type GroupKey = 'logs' | 'metrics';
type LokiGroupedRequest = Record<GroupKey, { request: DataQueryRequest<LokiQuery>; partition: TimeRange[] }>
```
```ts
export function getNextRequestPointers(requests: LokiGroupedRequest, requestKey: GroupKey, requestN: number):{nextRequestKey: GroupKey, nextRequestN: number} {
// replace `requestIndex === 0` with `requestKey === "logs"` and `requests[0]` with `requests["logs"]` etc.
}
```
}

what do you think?

Reply:
I've just take a second look at this, and currently did some small changes to improve the readability and cohesiveness of the code. After that:

> we assume the requests array's length is always 2.

The length can be 1 or 2, depending on query types: metric and logs, only logs, or only metrics.

```javascript
const requests = [];
  if (logQueries.length) {
    requests.push({
      request: { ...request, targets: logQueries },
      partition: partitionTimeRange(true, request.range, request.intervalMs, logQueries[0].resolution ?? 1),
    });
  }
  if (metricQueries.length) {
    requests.push({
      request: { ...request, targets: metricQueries },
      partition: partitionTimeRange(false, request.range, request.intervalMs, metricQueries[0].resolution ?? 1),
    });
  }
```

Additionally, I was thinking that it could be 3 in the near future when we add instant queries:

```javascript
const requests = [];
  if (logQueries.length) {
    // ...
  }
  if (metricQueries.length) {
    // ...
  }
  if (instantQueries.length) {
    requests.push({
      request: { ...request, targets: instantQueries },
      partition: request.range, // the original range, without partition
    });
  }
```

The reason why I used numerical indexes and not associative indexes was to be able to quickly iterate over the groups:

```javascript
function getNextRequestPointers(requests: LokiGroupedRequest, requestGroup: number, requestN: number) {
  // There's a pending request from the next group:
  if (requests[requestGroup + 1]?.partition[requestN - 1]) {
    return {
      nextRequestGroup: requestGroup + 1,
      nextRequestN: requestN,
    };
  }
  return {
    nextRequestGroup: 0,
    nextRequestN: requestN - 1,
  };
}
```

Without having to add harcoded checks to get the next one, like 

```javascript
// pseudocode
if (currentGroup === 'logs') then next group = 'metrics'
else if (currentGroup === 'metrics') then next group = 'instant'
else if (currentGroup === 'instant) then next group === 'logs'
// etc.
```

I don't mind spending more time on this, so feel free to share your thoughts. Hopefully, my intentions here are clearer now.

Reply:
hmm, i see your point. so, after the `LokiGroupedRequest` gets created, there is no need to special-case anything because every item in that array can be treated the same way. in that case i'm totally ok with keeping it as an array.